### PR TITLE
[AutoDiff] Generalize handling of semantic result parameters

### LINKED ISSUE #67174: Refine the implicit `wrt` logic for @differentiable attribute
_Originally posted by @rxwei in https://github.com/apple/swift/issues/66873#issuecomment-1610795910_

I think there are two angles where we need to decide:
- Whether non-wrt parameters should be treated as semantic results.
    - It seems that our conclusion was "no". 
- The default wrt inference behavior, i.e. the behavior when you apply `@differentiable` without specifying `wrt:`.
    - My current preference is to never infer wrt-ness on any `inout` argument. If the user explicitly specifies `wrt:` on an inout argument, we treat it as a semantic result.
    - Enforcing explicit wrt annotations has better clarity. We can always add inference rules in the future if there's a usability benefit, without breaking source code.

```swift
@differentiable // inferred as wrt: x, semantic result: formal result (no y)
func foo(x: Float, y: inout Float) -> Float

@differentiable // error: 'String' does not conform to 'Differentiable'
func foo(x: Float, y: inout Float) -> String

@differentiable // error: function does not have a differentiable return type
                // note: did you mean to differentiate wrt the inout parameter `y`?
                // fixit: insert `(wrt: y)`
func foo(x: Float, y: inout Float)
```

### PR DESCRIPTION
**Pull request by asl**
Introduce "semantic result parameters" abstraction and treat inouts via this abstraction only.
Do not consider non-wrt semantic result parameters as semantic results

Fixes https://github.com/apple/swift/issues/67174

### CODE REVIEW

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@rxwei gentle ping :)

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -3177,6 +3177,12 @@ class AnyFunctionType : public TypeBase {
     /// Whether the parameter is marked '@noDerivative'.
     bool isNoDerivative() const { return Flags.isNoDerivative(); }
 
+    /// Whether the parameter might be a semantic result. This includes
+    /// inout parameters
```
Comment (rxwei) [inline include/swift/AST/Types.h:3181]:
```suggestion
    /// inout parameters.
```

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -3177,6 +3177,12 @@ class AnyFunctionType : public TypeBase {
...
+    bool isSemanticResult() const {
```
Comment (rxwei) [inline include/swift/AST/Types.h:3182]: "Semantic result" is an AutoDiff term, isn't? Should it be named `isAutoDiffSemanticResult`?

Comment (asl) [reply include/swift/AST/Types.h:3182]: Looks reasonable, yes.

**Comment by asl** [comment]
@rxwei I renamed the methods outside of autodiff code. Also added 2 related commits from WIP class differentiation branch that are relevant for semantic result parameters. PTAL

Thanks!

**Review by rxwei — APPROVED** [review]

**Comment by asl** [comment]
@swift-ci please test

**Pull request merged**